tiledb/sm/storage_manager/storage_manager.cc

```diff
@@ -1926,6 +1928,12 @@ Status StorageManager::write(const URI& uri, void* data, uint64_t size) const {
   return vfs_->write(uri, data, size);
 }
 
+int StorageManager::num_threads() const {
+  if (num_threads_ < 0)
+    return DEFAULT_CONCURRENCY;
```

Comment:
This is already the default behavior for the configuration. Why is it re-implemented here?

Reply:
The `Config` class is storing strings. Is there another way to get the number of threads in `int` form, checking for the case of (default) `-1`?

Reply:
Oh wait, I see we're now overriding `-1`  with `std::thread::hardware_concurrency()`. Why not make this the default behavior in the config?

Reply:
That's when you don't have TBB:
```
#ifdef HAVE_TBB
  const int DEFAULT_CONCURRENCY =
      tbb::task_scheduler_init::default_num_threads();
#else
  const int DEFAULT_CONCURRENCY = std::thread::hardware_concurrency();
#endif
```

Happy to change this. The issue here is when `sm.num_tbb_threads` is `-1`. We parse this on-the-fly when allocating the TBB thread pool here:

```
Status init_tbb(const Config* config) {
  int nthreads;
  if (!config) {
    nthreads = std::strtol(Config::SM_NUM_TBB_THREADS.c_str(), nullptr, 10);
  } else {
    bool found = false;
    RETURN_NOT_OK(config->get<int>("sm.num_tbb_threads", &nthreads, &found));
    assert(found);
  }

  if (nthreads == tbb::task_scheduler_init::automatic) {
    nthreads = tbb::task_scheduler_init::default_num_threads();
  }

....
```

In the case we don't have TBB, the number of threads is `1` (whereas I think it should be equal to the hardware concurrency - we can change that later for the thread pool). I could set `DEFAULT_CONCURRENCY` to `1` for consistency for non TBB configurations though.

Reply:
The bottom line is that there is no way to get the number of threads allocated for CPU execution from the storage manager, and we'll need this in various places. So we do need a `StorageManager::num_threads` function. The question is how to handle the default `-1`, and the cleanest way is through some constant. I am open to placing that constant elsewhere (e.g., the `constants.{h.cc}` files).

Reply:
If we go with the current patch, we lose the functionality to disable completely concurrency in the `parallel_for` operations because we're now overriding it. If this is OK, let's just move forward to defaulting to the number of hardware threads everywhere instead of `1`. So I would say just change the default behavior in `config.h` and remove the `if (nthreads == tbb::task_scheduler_init::automatic)` path in `init_tbb` to simplify our thread count logic.

Reply:
I set `DEFAULT_CONCURRENCY` to 1 for the non-TBB compilation.